src/sendqueue.rs

```diff
@@ -0,0 +1,93 @@
+use std::convert::TryInto;
+use std::ptr::NonNull;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use libc::{c_int, c_uint};
+
+use crate::raw;
+use crate::Error;
+use crate::{Active, Capture};
+
+pub struct SendQueue {
+    squeue: NonNull<raw::pcap_send_queue>,
+    sync: c_int,
+}
+
+impl SendQueue {
+    pub fn new(memsize: c_uint) -> Result<Self, Error> {
+        let squeue = unsafe { raw::pcap_sendqueue_alloc(memsize) };
+        let squeue = if let Some(squeue) = NonNull::new(squeue) {
+            squeue
+        } else {
+            return Err(Error::InsufficientMemory);
+        };
+
+        Ok(Self { squeue, sync: 0 })
+    }
+
+    pub fn maxlen(&self) -> c_int {
+        unsafe { (*self.squeue.as_ptr()).maxlen }
+    }
+
+    pub fn len(&self) -> c_int {
+        unsafe { (*self.squeue.as_ptr()).len }
+    }
+
+    pub fn sync(&mut self, sync: bool) {
+        self.sync = if sync { 1 } else { 0 }
+    }
+
+    /// Add a packet to the queue.
+    pub fn queue(&mut self, buf: &[u8]) -> Result<(), Error> {
+        let start = SystemTime::now();
+        let since_the_epoch = start
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards");
+
+        let s = since_the_epoch.as_secs();
+        let us = since_the_epoch.subsec_micros();
+
+        let pkthdr = raw::pcap_pkthdr {
+            ts: libc::timeval {
+                tv_sec: s as i32,
+                tv_usec: us as i32,
+            },
+            caplen: buf.len() as u32,
+            len: buf.len() as u32,
+        };
+
+        let ph = &pkthdr as *const _;
+        let res = unsafe { raw::pcap_sendqueue_queue(self.squeue.as_ptr(), ph, buf.as_ptr()) };
+        if res == -1 {
+            return Err(Error::InsufficientMemory);
+        }
+
+        Ok(())
+    }
+
+    /// Transmit the contents of the queue.
+    pub fn transmit(&mut self, dev: &mut Capture<Active>) -> Result<(), Error> {
```

Comment:
why don't you mimic pcap api and let user define `sync` bool here ?

Reply:
- My original reasoning was that I would avoid even getting the time when a packet is added to the queue if `sync` is disabled.  This, however, did not happen -- but it could!
- From a usability point of view, if you know you need sync, you're going to know that you need to figure out how to set it.  If you don't know if you need need sync, it is extremely unlikely that you do, and that parameter will just add unnecessary confusion.  (I have conducted empirical experimentation on myself to come to this conclusion :)

I think this API design is better than the C one, but I can be convinced that it's better to mimic the original C API.